### ISSUE TITLE
[Feature request] Add fields

### ISSUE THREAD

**Issue by hnakamur** [pkg/errors#70]
First of all, thanks for creating and sharing a wonderful library!

I think a stack trace with an error message is a big help for debugging, but I also want variable values at the stack where the error occurred. I would like to log those values at the higher stack where I received the error.

So how about adding something like `NewWithFields` or `WrapWithFields`?

I implemented a proof of concept at [Comparing pkg:a2d6902c6d2a2f194eb3fb474981ab7867c81505...hnakamur:add_fields · pkg/errors](https://github.com/pkg/errors/compare/a2d6902c6d2a2f194eb3fb474981ab7867c81505...hnakamur:add_fields)

Could you tell me what do you think?
Thanks!

**Comment by hnakamur** [comment]
I read [(Gophercon 2016) Structured Logging and You, in 5 minute](https://docs.google.com/presentation/d/1632rG5GJD-bGqzaV4ILBnHEJ2mnw8rVbulp-LyKTtuY/edit#slide=id.gd0fe35de6_0_242) and [log/log.go at master · aybabtme/log](https://github.com/aybabtme/log/blob/master/log.go).
Now I think it would be better to use something like [github.com/go-kit/kit/log.Context](https://github.com/go-kit/kit/blob/1c5eda624c0266a464594db0acd64d741c7a5722/log/log.go#L71) which has ordered key-value pairs.

**Comment by davecheney** [comment]
Thank you for raising this feature request. I believe this is a duplicate of #34.

I believe I understand what you are asking for, but as I concluded in issue #34 such a feature could not be added without creating a mechanism for state to be passed via a side channel in the error's context. 

I do not disagree that this would not be useful, I'm more partial to this idea, moreso than structured logging, but I have decided to not implement this feature at this time as I am concerned that it would be abused.
